### PR TITLE
Fix dominance calculation

### PR DESCRIPTION
**Pull request by dneto0**
Fixes dominance calculation when there is a forward arc from an
unreachable block A to a reachable block B.  Before this fix, we would
say that B is not dominated by the graph entry node, and instead say
that the immediate dominator of B is the psuedo-entry node of the
augmented CFG.

The fix:

- Dominance is defined in terms of a traversal from the entry block
  of the CFG.  So the forward DFS should start from the function
  entry block, not the pseudo-entry-block.

- When following edges backward during dominance calculations, only go to
  nodes that are actually reachable in the forward traversal.
  Important: the sense of reachability flips around when computing
  post-dominance.

Fixes https://github.com/KhronosGroup/SPIRV-Tools/issues/297

### CODE REVIEW

**Comment by dneto0** [comment]
Rebased

**Comment by dneto0** [comment]
I'm interested in feedback from @umar456

File: test/Validate.CFG.cpp [hunk 1]
```diff
@@ -1149,6 +1149,41 @@ TEST_F(ValidateCFG, LoopWithoutBackEdgesBad) {
...
+  Block inner_merge("inner_merge");
+  Block exit("exit", SpvOpReturn);
+
+  entry.SetBody("%cond    = OpSLessThan %intt %one %two\n");
+  if (is_shader) {
+    entry.AppendBody("OpSelectionMerge %exit None\n");
+    inner_head.SetBody("OpSelectionMerge %inner_merge None\n");
+  }
+
+  string str = header(GetParam()) + nameOps("entry", "inner_merge", "exit") +
```
Comment (umar456) [inline test/Validate.CFG.cpp]: You don't need named Ids if you expect the module to validate successfully. I only add those when I need to validate the error message. Not a big deal.

Comment (dneto0) [inline test/Validate.CFG.cpp]: Ack.  It's also helpful to make sure Ids have numbers 1,2,3....
Good for debugging.  :-)

**Comment by umar456** [comment]
+2

**Pull request merged**